src/Migration/Sources/Csv.php

```diff
@@ -0,0 +1,338 @@
+                break;
+            }
+        }
+
+        $attributeTypes = [];
+        $manyToManyKeys = [];
+
+        foreach ($allAttributes as $attribute) {
+            $key = $attribute['key'];
+
+            if (
+                // Skip child-side relationships entirely
+                $attribute['type'] === Attribute::TYPE_RELATIONSHIP &&
+                ($attribute['side'] ?? '') === UtopiaDatabase::RELATION_SIDE_CHILD
+            ) {
+                continue;
+            }
+
+            $attributeTypes[$key] = $attribute['type'];
+
+            if (
+                $attribute['type'] === Attribute::TYPE_RELATIONSHIP &&
+                ($attribute['relationType'] ?? '') === 'manyToMany' &&
+                ($attribute['side'] ?? '') === 'parent'
+            ) {
+                $manyToManyKeys[] = $key;
+            }
+        }
+
+        $buffer = [];
+
+        while (($row = fgetcsv($stream)) !== false) {
+            $data = array_combine($headers, $row);
+            if ($data === false) {
+                continue;
+            }
+
+            $parsedData = $data;
+
+            foreach ($data as $key => $value) {
+                if (! isset($attributeTypes[$key])) {
+                    continue;
+                }
+
+                $type = $attributeTypes[$key];
+                $parsedValue = trim($value);
+
+                if ($parsedValue === '') {
+                    $parsedData[$key] = null;
+
+                    continue;
+                }
+
+                if (in_array($key, $manyToManyKeys, true)) {
+                    $parsedData[$key] = str_contains($parsedValue, ',')
+                        ? array_map('trim', explode(',', $parsedValue))
+                        : [$parsedValue];
+
+                    continue;
+                }
+
+                switch ($type) {
+                    case Attribute::TYPE_INTEGER:
+                        $parsedData[$key] = is_numeric($parsedValue) ? (int) $parsedValue : null;
+                        break;
+
+                    case Attribute::TYPE_FLOAT:
+                        $parsedData[$key] = is_numeric($parsedValue) ? (float) $parsedValue : null;
+                        break;
+
+                    case Attribute::TYPE_BOOLEAN:
+                        $parsedData[$key] = filter_var($parsedValue, FILTER_VALIDATE_BOOLEAN);
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+
+            $permissions = [];
+            $documentId = $parsedData['$id'] ?? 'unique()';
+
+            if ($hasDocumentSecurityEnabled && isset($parsedData['$permissions'])) {
+                $permissions = $this->parsePermissions($parsedData['$permissions']);
+            }
+
+            foreach ($parsedData as $key => $value) {
+                if (str_starts_with($key, '$')) {
+                    unset($parsedData[$key]);
+                }
+            }
```

Comment:
We can remove to reduce looping, the DB will do this validation for us

Reply:
I had a csv with `$permissions`, it would throw error because then -
1. The parsedData has a column
2. We also passed $permissions that are parsed.

We could do a clean `unset` for id and permissions.